### PR TITLE
feat(searchbar): add cancel button options

### PR DESCRIPTION
**Pull request by liamdebeasi**
## Pull request checklist

Please check if your PR fulfills the following requirements:
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] Build (`npm run build`) was run locally and any changes were pushed
- [x] Lint (`npm run lint`) has passed locally and any fixes were made for failures

## Pull request type

 

 

Please check the type of change your PR introduces:
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe): 

## What is the current behavior?

Issue Number: https://github.com/ionic-team/ionic/issues/18399

## What is the new behavior?

- Adds `always`, `never`, and `focus` options for `show-cancel-button`
- Deprecates  the `true` and `false` options

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Other information

Original PR: https://github.com/ionic-team/ionic/pull/18401

### CODE REVIEW

File: core/src/components/searchbar/searchbar.tsx [hunk 1]
```diff
@@ -19,7 +19,7 @@ export class Searchbar implements ComponentInterface {
   private isCancelVisible = false;
   private shouldAlignLeft = true;
 
-  @Element() el!: HTMLElement;
+  @Element() el!: HTMLStencilElement;
```
Comment (manucorporat) [inline core/src/components/searchbar/searchbar.tsx]:
```suggestion
  @Element() el!: HTMLIonSearchbarElement;
```

File: core/src/components/searchbar/searchbar.tsx [hunk 2]
```diff
@@ -362,6 +374,54 @@ export class Searchbar implements ComponentInterface {
...
+  }
+
+  /**
+   * Wrapper method to check if the cancel button
+   * should be shown on focus.
+   *
+   * TODO: Remove this when the `true` and `false`
+   * options are removed.
+   */
+  private isCancelButtonSetToFocus(): boolean {
```
Comment (manucorporat) [inline core/src/components/searchbar/searchbar.tsx]: can we make it a pure function that takes the value? will minify much better:

```js
const isCancelButtonSetToFocus = (showCancelButton: boolean | string) => (
      this.showCancelButton === 'focus' ||
      this.showCancelButton === 'true' ||
      this.showCancelButton === true ||
      this.showCancelButton === ''
);
```

File: core/src/components/searchbar/searchbar.tsx [hunk 3]
```diff
@@ -362,6 +374,54 @@ export class Searchbar implements ComponentInterface {
...
+  }
+
+  /**
+   * Wrapper method to check if the cancel button
+   * should never be shown.
+   *
+   * TODO: Remove this when the `true` and `false`
+   * options are removed.
+   */
+  private isCancelButtonSetToNever(): boolean {
```
Comment (manucorporat) [inline core/src/components/searchbar/searchbar.tsx]: same, pure function

File: core/src/components/searchbar/searchbar.tsx [hunk 4]
```diff
@@ -101,9 +103,13 @@ export class Searchbar implements ComponentInterface {
...
   /**
-   * If `true`, show the cancel button.
+   * Sets the behavior for the cancel button.
+   * Setting to `"focus"` shows the cancel button on focus.
+   * Setting to `"never"` hides the cancel button.
+   * Setting to `"always"` shows the cancel button regardless
+   * of focus state.
    */
-  @Prop() showCancelButton = false;
+  @Prop() showCancelButton: boolean | string = false;
```
Comment (brandyscarney) [inline core/src/components/searchbar/searchbar.tsx]: Should we change the default on this to "never" or undefined?

**Comment by brandyscarney** [comment]
Looks great, nice work!

**Pull request merged**